### PR TITLE
fixups for list.files on Windows

### PR DESCRIPTION
**Pull request by kevinushey**
### Intent

Addresses https://github.com/rstudio/rstudio/issues/10679.
Addresses https://github.com/rstudio/rstudio/issues/10672.

### Approach

Use R's own `grepl()` rather than `boost::regex`, to avoid incompatibilities between the two regular expression engines.

Also make sure we expand tilde-prefixed paths before attempting to list files. 😭 

### Automated Tests

Tests included and updated.

### QA Notes

Test via notes in https://github.com/rstudio/rstudio/issues/10679.

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

File: src/cpp/session/modules/SessionFiles.cpp [hunk 1]
```diff
@@ -1391,26 +1391,35 @@ class ListFilesAcceptAll
...
+      preserver_.add(callSEXP_);
    }
    
    bool operator()(const boost::filesystem::path& path)
    {
-      return boost::regex_search(path.native(), pattern_);
+      // construct call
+      r::sexp::Protect protect;
+      SEXP xSEXP = r::sexp::create(path.string(), &protect);
+      SETCADDR(callSEXP_, xSEXP);
```
Comment (jgutman) [inline src/cpp/session/modules/SessionFiles.cpp:1413]: does this just set xSEXP to the second argument because that one happens to be R_NilValue?

**Review by jgutman — APPROVED** [review]
LGTM!

File: src/cpp/session/modules/SessionFiles.cpp (hunk 1 above)
Comment (kevinushey) [reply src/cpp/session/modules/SessionFiles.cpp:1413]: That's right. The R call object is basically like a list of elements ["grepl", <pattern>, <x>, <ignore.case>]; to set the 3rd element in that list (the second argument), we use `SETCADDR()`.

**Comment by kevinushey** [comment]
Holding on merge as we may want to take for PT and then merge back to SW.

**Pull request merged**